sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/SdkMeter.java

```diff
@@ -60,4 +69,44 @@ public DoubleHistogramBuilder histogramBuilder(String name) {
   public DoubleGaugeBuilder gaugeBuilder(String name) {
     return new SdkDoubleGaugeBuilder(meterProviderSharedState, meterSharedState, name);
   }
+
+  @Override
+  public BatchRecorder batch() {
+    return new SdkBatchRecorder(batchLatch);
+  }
+
+  static class BatchLatch {
```

Comment:
This is the naive synchronization mechanism that ensures the recordings are included in a collection atomically. 

When `BatchRecorder#record(Attributes, Context)` is called, `startBatchRecord()` is called on start. This blocks if a collection is in progress. If not, it acquires a permit from `batchSemaphore` that blocks collections until the record is completed and `finishBatchRecord()` is called to release the permit.

When a collect occurs, `startCollect()` is called. This sets `collectLatch = new CountDownLatch(1)`, which blocks batch records from occurring until  the collect completes. It then blocks while acquiring all the permits from `batchSemaphore`, which allows in progress recordings to complete. When the collect is done, it calls `finishCollect()`, and counts down the latch and returns the permits.